swarm/src/behaviour/external_addresses.rs

```diff
@@ -29,24 +18,60 @@ impl ExternalAddresses {
         self.addresses.iter()
     }
 
+    pub fn as_slice(&self) -> &[Multiaddr] {
+        self.addresses.as_slice()
+    }
+
     /// Feed a [`FromSwarm`] event to this struct.
     ///
     /// Returns whether the event changed our set of external addresses.
     pub fn on_swarm_event<THandler>(&mut self, event: &FromSwarm<THandler>) -> bool {
         match event {
             FromSwarm::ExternalAddrConfirmed(ExternalAddrConfirmed { addr }) => {
-                if self.addresses.len() < self.limit {
-                    return self.addresses.insert((*addr).clone());
+                if let Some(pos) = self
+                    .addresses
+                    .iter()
+                    .position(|candidate| candidate == *addr)
+                {
+                    // Refresh the existing confirmed address.
+                    self.addresses.remove(pos);
+                    self.push_front(addr);
+
+                    return false; // No changes to our external addresses.
+                }
+
+                self.push_front(addr);
+
+                if self.addresses.len() > MAX_LOCAL_EXTERNAL_ADDRS {
+                    self.addresses.pop();
                 }
```

Comment:
I still think this needs a log line. Otherwise I would be surprised if someone can easily debug why their previously confirmed external address is no longer part of the set. Any objections?

Reply:
Previously, we didn't have a log line either so I wouldn't consider this a regression.

Happy to add it in a follow-up!

Reply:
I guess you are saying that previously we just stopped adding them and now we remove them from the set, I see.

Will add a log line.